README.md
mongodb_store
==============

## Release Build status: 
* `ros-hydro-mongodb-log_binarydeb_precise_amd64` [![Build Status](http://lcas.lincoln.ac.uk/jenkins/job/ros-hydro-mongodb-log_binarydeb_precise_amd64/badge/icon)](http://lcas.lincoln.ac.uk/jenkins/job/ros-hydro-mongodb-log_binarydeb_precise_amd64)
* `ros-hydro-mongodb-store-cpp-client_binarydeb_precise_amd64` [![Build Status](http://lcas.lincoln.ac.uk/jenkins/job/ros-hydro-mongodb-store-cpp-client_binarydeb_precise_amd64/badge/icon)](http://lcas.lincoln.ac.uk/jenkins/job/ros-hydro-mongodb-store-cpp-client_binarydeb_precise_amd64)
* `ros-hydro-mongodb-store-msgs_binarydeb_precise_amd64` [![Build Status](http://lcas.lincoln.ac.uk/jenkins/job/ros-hydro-mongodb-store-msgs_binarydeb_precise_amd64/badge/icon)](http://lcas.lincoln.ac.uk/jenkins/job/ros-hydro-mongodb-store-msgs_binarydeb_precise_amd64)
* `ros-hydro-mongodb-store_binarydeb_precise_amd64` [![Build Status](http://lcas.lincoln.ac.uk/jenkins/job/ros-hydro-mongodb-store_binarydeb_precise_amd64/badge/icon)](http://lcas.lincoln.ac.uk/jenkins/job/ros-hydro-mongodb-store_binarydeb_precise_amd64)


mongodb_log/CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package mongodb_log
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.6.0 (2022-09-20)
------------------
* Ensure Python 3 compatibility in mongodb log (`#277 <https://github.com/strands-project/mongodb_store/issues/277>`_)
* update package.xml to format=3 (`#269 <https://github.com/strands-project/mongodb_store/issues/269>`_)
* Contributors: Gal Gorjup, Kei Okada, Nick Hawes

0.5.2 (2019-11-11)
------------------
* back to system mongo
* Contributors: Ferenc Balint-Benczedi, Nick Hawes, Shingo Kitagawa, Volker Gabler

0.5.1 (2019-06-28)
------------------

0.5.0 (2018-12-20)
------------------
* Merge pull request `#231 <https://github.com/strands-project/mongodb_store/issues/231>`_ from bbferka/melodic-devel
  Melodic devel
* back to system mongo
* Contributors: Ferenc Balint-Benczedi, Nick Hawes

0.4.5 (2019-06-28)
------------------

0.4.3 (2019-04-04)
------------------
* Merge pull request `#238 <https://github.com/strands-project/mongodb_store/issues/238>`_ from charlielito/patch-1
  Fix typo in log message
* Fix typo in log message
  It is subscribe rather than subsribe
* Contributors: Carlos Andrés Álvarez Restrepo, Nick Hawes

0.4.2 (2018-06-04)
------------------
* instead of lsb-release read /etc/os-version from CMake to find OS version
* Contributors: Ferenc Balint-Benczedi

0.4.1 (2018-05-29)
------------------
* check if env. var. exists if not decide based on OS version
* Contributors: Ferenc Balint-Benczedi

0.4.0 (2018-05-23)
------------------
* if compiler supports it use C++11
* Contributors: Ferenc Balint-Benczedi

0.3.8 (2018-05-02)
------------------

0.3.7 (2017-12-14)
------------------
* Merge pull request `#210 <https://github.com/strands-project/mongodb_store/issues/210>`_ from hawesie/kinetic-devel
  Updated to build on OS X with the mongo client legacy version.
* Used the openssl libs from find_package.
* Updated to build on OS X with the mongo client legacy version.
  This makes a couple of changes. For libmongocxx_ros the paths for openssl are added to the scons command and all ".so" and set to ".so" or ".dylib" depending on platform. For mongodb_log this removes the explicit linking of mongoclient which isn't needed since catkin_libraries includes the lib created by libmongocxx_ros.
* Contributors: Nick Hawes

0.3.6 (2017-09-14)
------------------

0.3.5 (2017-09-03)
------------------

0.3.4 (2017-09-03)
------------------

0.3.3 (2017-09-02)
------------------

0.3.2 (2017-08-31)
------------------

0.3.1 (2017-08-24)
------------------


0.3.0 (2017-08-24)
------------------


0.2.2 (2017-06-28)
------------------

0.2.1 (2017-06-28)
------------------

0.2.0 (2017-06-28)
------------------
* dependencies fixing
* Contributors: Marc Hanheide, Nick Hawes

0.1.30 (2017-06-23)
-------------------

0.1.29 (2017-06-19)
-------------------

0.1.28 (2016-11-09)
-------------------

0.1.27 (2016-11-01)
-------------------

0.1.26 (2016-10-14)
-------------------
* indigo-0.1.25
* Updating changelogs
* indigo-0.1.24
* updated Changelogs
* Contributors: Nick Hawes

0.1.25 (2016-04-28)
-------------------

0.1.24 (2016-04-19)
-------------------

0.1.23 (2016-04-19)
-------------------
* Fixed missing namespace for mongodb_log
* indigo-0.1.22
* updated Changelogs
* Contributors: Nick Hawes


0.1.22 (2016-02-23)
-------------------


0.1.20 (2015-11-11)
-------------------



0.1.19 (2015-10-28)
-------------------

0.1.18 (2015-10-28)
-------------------

0.1.17 (2015-09-01)
-------------------

0.1.16 (2015-08-04)
-------------------
* add option to treat topic name arguments as regular expression
* add option to specify collection name
* Contributors: Furushchev

0.1.15 (2015-05-10)
-------------------

0.1.14 (2015-04-27)
-------------------
* Adding install targets for mongodb_log
  Closes `#129 <https://github.com/strands-project/mongodb_store/issues/129>`_
* Contributors: Christian Dondrup

0.1.13 (2015-04-22)
-------------------
* Recheck topics at a fixed interval in order to attempt to resubscribe to topics that were missing at startup.
  This closes `#126 <https://github.com/strands-project/mongodb_store/issues/126>`_.
* Changed mongodb_log to not wait for topics to be published, instead subscribing to all the other topics
* Contributors: Nick Hawes, Nils Bore

0.1.12 (2015-02-09)
-------------------

0.1.11 (2015-02-09)
-------------------
* Extended usage output string by new command line options
* Changed default behaviour back to its former way
  The 'a' command line parameter now activates throttling; not specifying it makes the logger log all tf transformations.
* Added throttling capabilities for high-frequency tf logging
  Added tf logging throttling capabilities originally introduced in https://github.com/code-iai/ros-mongodb_log. A transform is only logged when either this transform has not been logged before, or when the new version of this transform is sufficiently different from the one logged before. Additional command line parameters can be used to control how throttling is done:
  * `-k <d>`: Cartesian (vectorial) distance (in meters) threshold between the old and the new transform
  * `-l <d>`: Angular diastance (in deg) threshold between the old and the new transform
  * `-g <d>`: At least log every transform every `d` seconds, even if nothing changed
  * `-a`: Always log, don't throttle
* Contributors: Jan Winkler

0.1.10 (2014-11-23)
-------------------

0.1.9 (2014-11-18)
------------------
* Use rospy to remove additional arguments when launched via roslaunch
* Contributors: Nils Bore

0.1.8 (2014-11-11)
------------------

0.1.7 (2014-11-09)
------------------

0.1.6 (2014-11-06)
------------------

0.1.5 (2014-11-05)
------------------

0.1.4 (2014-10-29)
------------------

0.1.3 (2014-10-21)
------------------

0.1.2 (2014-10-20)
------------------

0.1.1 (2014-10-17)
------------------

0.1.0 (2014-10-16)
------------------

src/mongodb_rrd.py
#!/usr/bin/python

###########################################################################
#  mongorrd.py - create RRD graphs for MongoDB using rrdtool
#
#  Created: Tue Nov 23 09:29:37 2010
#  Copyright  2010  Tim Niemueller [www.niemueller.de]
#                   Carnegie Mellon University
#                   Intel Labs Pittsburgh
###########################################################################

#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU Library General Public License for more details.
#
#  Read the full text in the LICENSE.GPL file in the doc directory.

# make sure we aren't using floor division
from __future__ import division, with_statement

import os
import sys
import random
from pprint import pprint
from optparse import OptionParser
from datetime import datetime, timedelta
from time import sleep

from pymongo import Connection
import rrdtool

LOOPTIME  = 10
GRAPHTIME = 60

_quit    = False
_conn    = None
_admindb = None
_datadb  = None

class MongoRRD(object):
    def __init__(self, graph_clear = False, mongodb_host=None, mongodb_port=None, mongodb_name="roslog"):
        self.quit        = False
        self.graph_clear = graph_clear
        self._conn       = Connection(mongodb_host, mongodb_port)
        self._admindb    = self._conn["admin"]
        self._datadb     = self._conn[mongodb_name]

        self.init_rrd()

    def create_rrd(self, file, *data_sources):
        if not os.path.isfile(file) or self.graph_clear:
            rrdtool.create(file, "--step", "10", "--start", "0",
                           # remember that we always need to add the previous RRA time range
                           # hence number of rows is not directly calculated by desired time frame
                           "RRA:AVERAGE:0.5:1:720",    #  2 hours of 10 sec  averages
                           "RRA:AVERAGE:0.5:3:1680",   # 12 hours of 30 sec  averages
                           "RRA:AVERAGE:0.5:30:456",   #  1 day   of  5 min  averages
                           "RRA:AVERAGE:0.5:180:412",  #  7 days  of 30 min  averages
                           "RRA:AVERAGE:0.5:720:439",  #  4 weeks of  2 hour averages
                           "RRA:AVERAGE:0.5:8640:402", #  1 year  of  1 day averages
                           "RRA:MIN:0.5:1:720",
                           "RRA:MIN:0.5:3:1680",
                           "RRA:MIN:0.5:30:456",
                           "RRA:MIN:0.5:180:412",
                           "RRA:MIN:0.5:720:439",
                           "RRA:MIN:0.5:8640:402",
                           "RRA:MAX:0.5:1:720",
                           "RRA:MAX:0.5:3:1680",
                           "RRA:MAX:0.5:30:456",
                           "RRA:MAX:0.5:180:412",
                           "RRA:MAX:0.5:720:439",
                           "RRA:MAX:0.5:8640:402",
                           *data_sources)
        

    def init_rrd(self):
        self.create_rrd("opcounters.rrd",
                        "DS:insert:COUNTER:30:0:U",
                        "DS:query:COUNTER:30:0:U",
                        "DS:update:COUNTER:30:0:U",
                        "DS:delete:COUNTER:30:0:U",
                        "DS:getmore:COUNTER:30:0:U",
                        "DS:command:COUNTER:30:0:U")

        self.create_rrd("memory.rrd",
                        "DS:resident:GAUGE:30:0:U",
                        "DS:virtual:GAUGE:30:0:U",
                        "DS:mapped:GAUGE:30:0:U")
        
        self.create_rrd("indexes.rrd",
                        "DS:accesses:COUNTER:30:0:U",
                        "DS:hits:COUNTER:30:0:U",
                        "DS:misses:COUNTER:30:0:U",
                        "DS:resets:COUNTER:30:0:U")

        self.create_rrd("locks.rrd", "DS:locktime:COUNTER:30:0:U")

        self.create_rrd("dbstats.rrd",
                        "DS:collections:GAUGE:30:0:U",
                        "DS:objects:GAUGE:30:0:U",
                        "DS:avgObjSize:GAUGE:30:0:U",
                        "DS:dataSize:GAUGE:30:0:U",
                        "DS:storageSize:GAUGE:30:0:U",
                        "DS:numExtents:GAUGE:30:0:U",
                        "DS:indexes:GAUGE:30:0:U",
                        "DS:indexSize:GAUGE:30:0:U",
                        "DS:fileSize:GAUGE:30:0:U")

    def graph_rrd(self):
        rrdtool.graph("opcounters.png",
                      "--start=-3600", "--end=-10",
                      "--disable-rrdtool-tag", "--width=560",
                      "--font", "LEGEND:10:", "--font", "UNIT:8:",
                      "--font", "TITLE:12:", "--font", "AXIS:8:",
                      "--title=MongoDB Op Counters",
                      "DEF:insert=opcounters.rrd:insert:AVERAGE:step=10",
                      "DEF:query=opcounters.rrd:query:AVERAGE:step=10",
                      "DEF:update=opcounters.rrd:update:AVERAGE:step=10",
                      "DEF:delete=opcounters.rrd:delete:AVERAGE:step=10",
                      "DEF:getmore=opcounters.rrd:getmore:AVERAGE:step=10",
                      "DEF:command=opcounters.rrd:command:AVERAGE:step=10",
                      "LINE1:insert#FF7200:Inserts",
                      "GPRINT:insert:LAST: Current\\:%8.2lf %s",
                      "GPRINT:insert:AVERAGE:Average\\:%8.2lf %s",
                      "GPRINT:insert:MAX:Maximum\\:%8.2lf %s\\n",
                      "LINE1:query#503001:Queries",
                      "GPRINT:query:LAST: Current\\:%8.2lf %s",
                      "GPRINT:query:AVERAGE:Average\\:%8.2lf %s",
                      "GPRINT:query:MAX:Maximum\\:%8.2lf %s\\n",
                      "LINE1:update#EDAC00:Updates",
                      "GPRINT:update:LAST: Current\\:%8.2lf %s",
                      "GPRINT:update:AVERAGE:Average\\:%8.2lf %s",
                      "GPRINT:update:MAX:Maximum\\:%8.2lf %s\\n",
                      "LINE1:delete#506101:Deletes",
                      "GPRINT:delete:LAST: Current\\:%8.2lf %s",
                      "GPRINT:delete:AVERAGE:Average\\:%8.2lf %s",
                      "GPRINT:delete:MAX:Maximum\\:%8.2lf %s\\n",
                      "LINE1:getmore#0CCCCC:Getmores",
                      "GPRINT:getmore:LAST:Current\\:%8.2lf %s",
                      "GPRINT:getmore:AVERAGE:Average\\:%8.2lf %s",
                      "GPRINT:getmore:MAX:Maximum\\:%8.2lf %s\\n",
                      "LINE1:command#53CA05:Commands",
                      "GPRINT:command:LAST:Current\\:%8.2lf %s",
                      "GPRINT:command:AVERAGE:Average\\:%8.2lf %s",
                      "GPRINT:command:MAX:Maximum\\:%8.2lf %s\\n")

        rrdtool.graph("memory.png",
                      "--start=-3600", "--end=-10",
                      "--disable-rrdtool-tag",
                      "--width=560",
                      "--font", "LEGEND:10:", "--font", "UNIT:8:",
                      "--font", "TITLE:12:", "--font", "AXIS:8:",
                      "--title=MongoDB Memory Usage",
                      "DEF:rawresident=memory.rrd:resident:AVERAGE:step=10",
                      "DEF:rawvirtual=memory.rrd:virtual:AVERAGE:step=10",
                      "DEF:rawmapped=memory.rrd:mapped:AVERAGE:step=10",
                      "CDEF:resident=rawresident,1048576,*",
                      "CDEF:virtual=rawvirtual,1048576,*",
                      "CDEF:mapped=rawmapped,1048576,*",
                      "AREA:virtual#3B7AD9:Virtual",
                      "GPRINT:virtual:LAST: Current\\:%8.2lf %s",
                      "GPRINT:virtual:AVERAGE:Average\\:%8.2lf %s",
                      "GPRINT:virtual:MAX:Maximum\\:%8.2lf %s\\n",
                      "AREA:mapped#6FD1BF:Mapped",
                      "GPRINT:mapped:LAST:  Current\\:%8.2lf %s",
                      "GPRINT:mapped:AVERAGE:Average\\:%8.2lf %s",
                      "GPRINT:mapped:MAX:Maximum\\:%8.2lf %s\\n",
                      "AREA:resident#0E6E5C:Resident",
                      "GPRINT:resident:LAST:Current\\:%8.2lf %s",
                      "GPRINT:resident:AVERAGE:Average\\:%8.2lf %s",
                      "GPRINT:resident:MAX:Maximum\\:%8.2lf %s\\n")
        
        rrdtool.graph("dbstats_collindext.png",
                      "--start=-3600", "--end=-10",
                      "--disable-rrdtool-tag",
                      "--width=560",
                      "--font", "LEGEND:10:", "--font", "UNIT:8:",
                      "--font", "TITLE:12:", "--font", "AXIS:8:",
                      "--title=MongoDB Stats (Collections, Indexes, Extents)",
                      "DEF:collections=dbstats.rrd:collections:AVERAGE:step=1",
                      "DEF:indexes=dbstats.rrd:indexes:AVERAGE:step=1",
                      "DEF:numExtents=dbstats.rrd:numExtents:AVERAGE:step=1",
                      "LINE1:collections#FF7200:Collections",
                      "GPRINT:collections:LAST:Current\\:%8.2lf %s",
                      "GPRINT:collections:AVERAGE:Average\\:%8.2lf %s",
                      "GPRINT:collections:MAX:Maximum\\:%8.2lf %s\\n",
                      "LINE1:indexes#EDAC00:Indexes",
                      "GPRINT:indexes:LAST:    Current\\:%8.2lf %s",
                      "GPRINT:indexes:AVERAGE:Average\\:%8.2lf %s",
                      "GPRINT:indexes:MAX:Maximum\\:%8.2lf %s\\n",
                      "LINE1:numExtents#506101:Extents",
                      "GPRINT:numExtents:LAST:    Current\\:%8.2lf %s",
                      "GPRINT:numExtents:AVERAGE:Average\\:%8.2lf %s",
                      "GPRINT:numExtents:MAX:Maximum\\:%8.2lf %s\\n")

        rrdtool.graph("dbstats_objects.png",
                      "--start=-3600", "--end=-10",
                      "--disable-rrdtool-tag",
                      "--width=560",
                      "--font", "LEGEND:10:", "--font", "UNIT:8:",
                      "--font", "TITLE:12:", "--font", "AXIS:8:",
                      "--title=MongoDB Objects",
                      "DEF:objects=dbstats.rrd:objects:AVERAGE:step=1",
                      "LINE1:objects#503001:Objects",
                      "GPRINT:objects:LAST: Current\\:%8.2lf %s",
                      "GPRINT:objects:AVERAGE:Average\\:%8.2lf %s",
                      "GPRINT:objects:MAX:Maximum\\:%8.2lf %s\\n")

        rrdtool.graph("dbstats_sizes.png",
                      "--start=-3600", "--end=-10",
                      "--disable-rrdtool-tag",
                      "--width=560",
                      "--font", "LEGEND:10:", "--font", "UNIT:8:",
                      "--font", "TITLE:12:", "--font", "AXIS:8:",
                      "--title=MongoDB Stats (Sizes)",
                      "DEF:avgObjSize=dbstats.rrd:avgObjSize:AVERAGE:step=1",
                      "DEF:dataSize=dbstats.rrd:dataSize:AVERAGE:step=1",
                      "DEF:storageSize=dbstats.rrd:storageSize:AVERAGE:step=1",
                      "DEF:indexSize=dbstats.rrd:indexSize:AVERAGE:step=1",
                      "DEF:fileSize=dbstats.rrd:fileSize:AVERAGE:step=1",
                      "LINE1:avgObjSize#FF7200:Avg Object",
                      "GPRINT:avgObjSize:LAST:Current\\:%8.2lf %s",
                      "GPRINT:avgObjSize:AVERAGE:Average\\:%8.2lf %s",
                      "GPRINT:avgObjSize:MAX:Maximum\\:%8.2lf %s\\n",
                      "LINE1:dataSize#503001:Data",
                      "GPRINT:dataSize:LAST:      Current\\:%8.2lf %s",
                      "GPRINT:dataSize:AVERAGE:Average\\:%8.2lf %s",
                      "GPRINT:dataSize:MAX:Maximum\\:%8.2lf %s\\n",
                      "LINE1:storageSize#EDAC00:Storage",
                      "GPRINT:storageSize:LAST:   Current\\:%8.2lf %s",
                      "GPRINT:storageSize:AVERAGE:Average\\:%8.2lf %s",
                      "GPRINT:storageSize:MAX:Maximum\\:%8.2lf %s\\n",
                      "LINE1:indexSize#506101:Index",
                      "GPRINT:indexSize:LAST:     Current\\:%8.2lf %s",
                      "GPRINT:indexSize:AVERAGE:Average\\:%8.2lf %s",
                      "GPRINT:indexSize:MAX:Maximum\\:%8.2lf %s\\n",
                      "LINE1:fileSize#0CCCCC:File",
                      "GPRINT:fileSize:LAST:      Current\\:%8.2lf %s",
                      "GPRINT:fileSize:AVERAGE:Average\\:%8.2lf %s",
                      "GPRINT:fileSize:MAX:Maximum\\:%8.2lf %s\\n")

        rrdtool.graph("indexes.png",
                      "--start=-3600", "--end=-10",
                      "--disable-rrdtool-tag",
                      "--width=560",
                      "--font", "LEGEND:10:", "--font", "UNIT:8:",
                      "--font", "TITLE:12:", "--font", "AXIS:8:",
                      "--title=MongoDB Indexes",
                      "DEF:accesses=indexes.rrd:accesses:AVERAGE:step=1",
                      "DEF:hits=indexes.rrd:hits:AVERAGE:step=1",
                      "DEF:misses=indexes.rrd:misses:AVERAGE:step=1",
                      "DEF:resets=indexes.rrd:resets:AVERAGE:step=1",
                      "LINE1:accesses#FF7200:Accesses",
                      "GPRINT:accesses:LAST:Current\\:%8.2lf %s",
                      "GPRINT:accesses:AVERAGE:Average\\:%8.2lf %s",
                      "GPRINT:accesses:MAX:Maximum\\:%8.2lf %s\\n",
                      "LINE1:hits#503001:Hits",
                      "GPRINT:hits:LAST:    Current\\:%8.2lf %s",
                      "GPRINT:hits:AVERAGE:Average\\:%8.2lf %s",
                      "GPRINT:hits:MAX:Maximum\\:%8.2lf %s\\n",
                      "LINE1:misses#EDAC00:Misses",
                      "GPRINT:misses:LAST:  Current\\:%8.2lf %s",
                      "GPRINT:misses:AVERAGE:Average\\:%8.2lf %s",
                      "GPRINT:misses:MAX:Maximum\\:%8.2lf %s\\n",
                      "LINE1:resets#506101:Resets",
                      "GPRINT:resets:LAST:  Current\\:%8.2lf %s",
                      "GPRINT:resets:AVERAGE:Average\\:%8.2lf %s",
                      "GPRINT:resets:MAX:Maximum\\:%8.2lf %s\\n")

    
    def update_rrd(self):
        status = self._admindb.command("serverStatus")
        ocs = status["opcounters"]
        opcounters = (ocs["insert"], ocs["query"], ocs["update"], ocs["delete"], ocs["getmore"], ocs["command"])
        mem = status["mem"]
        meminfo = (mem["resident"], mem["virtual"], mem["mapped"])
        ind = status["indexCounters"]["btree"]
        indexinfo = (ind["accesses"], ind["hits"], ind["misses"], ind["resets"])
        lockinfo = status["globalLock"]["lockTime"]
        rrdtool.update("opcounters.rrd", "N:%d:%d:%d:%d:%d:%d" % opcounters)
        rrdtool.update("memory.rrd", "N:%d:%d:%d" % meminfo)
        rrdtool.update("indexes.rrd", "N:%d:%d:%d:%d" % indexinfo)
        rrdtool.update("locks.rrd", "N:%d" % lockinfo)

        dbstats = self._datadb.command("dbStats")
        dbs = (dbstats["collections"], dbstats["objects"], dbstats["avgObjSize"],
               dbstats["dataSize"], dbstats["storageSize"], dbstats["numExtents"],
               dbstats["indexes"], dbstats["indexSize"], dbstats["fileSize"])
        rrdtool.update("dbstats.rrd", "N:%d:%d:%f:%d:%d:%d:%d:%d:%d" % dbs)

    def run(self):
        looping_threshold  = timedelta(0, LOOPTIME,  0)
        graphing_threshold = timedelta(0, GRAPHTIME, 0)

        graphing_last = datetime.now()
        
        while not self.quit:
            started = datetime.now()

            self.update_rrd()

            if datetime.now() - graphing_last > graphing_threshold:
                print("Generating graphs")
                self.graph_rrd()
                graphing_last = datetime.now()

            # the following code makes sure we run once per second, taking
            # varying run-times and interrupted sleeps into account
            td = datetime.now() - started
            while td < looping_threshold:
                sleeptime = LOOPTIME - (td.microseconds + (td.seconds + td.days * 24 * 3600) * 10**6) / 10**6
                if sleeptime > 0:
                    sleep(sleeptime)
                td = datetime.now() - started

def main(argv):
    parser = OptionParser()
    parser.add_option("--mongodb-host", dest="mongodb_host",
                      help="Hostname of MongoDB", metavar="HOST",
                      default="localhost")
    parser.add_option("--mongodb-port", dest="mongodb_port",
                      help="Hostname of MongoDB", type="int",
                      metavar="PORT", default=27017)
    parser.add_option("--mongodb-name", dest="mongodb_name",
                      help="Name of DB in which to store values",
                      metavar="NAME", default="roslog")
    parser.add_option("--graph-clear", dest="graph_clear", default=False,
                      action="store_true",
                      help="Remove existing RRD files.")
    (options, args) = parser.parse_args()

    mongorrd = MongoRRD(graph_clear=options.graph_clear,
                        mongodb_host=options.mongodb_host,
                        mongodb_port=options.mongodb_port,
                        mongodb_name=options.mongodb_name)
    mongorrd.run()


if __name__ == "__main__":
    main(sys.argv)
